crates/ty_python_semantic/src/types/function.rs

```diff
@@ -0,0 +1,983 @@
+        let node = function_scope.node(db);
+        let func_def = node.as_function()?;
+        let range = parameter_index
+            .and_then(|parameter_index| {
+                func_def
+                    .parameters
+                    .iter()
+                    .nth(parameter_index)
+                    .map(|param| param.range())
+            })
+            .unwrap_or(func_def.parameters.range);
+        let name_span = span.clone().with_range(func_def.name.range);
+        let parameter_span = span.with_range(range);
+        Some((name_span, parameter_span))
+    }
+
+    pub(crate) fn spans(self, db: &'db dyn Db) -> Option<FunctionSpans> {
+        let function_scope = self.body_scope(db);
+        let span = Span::from(function_scope.file(db));
+        let node = function_scope.node(db);
+        let func_def = node.as_function()?;
+        let return_type_range = func_def.returns.as_ref().map(|returns| returns.range());
+        let mut signature = func_def.name.range.cover(func_def.parameters.range);
+        if let Some(return_type_range) = return_type_range {
+            signature = signature.cover(return_type_range);
+        }
+        Some(FunctionSpans {
+            signature: span.clone().with_range(signature),
+            name: span.clone().with_range(func_def.name.range),
+            parameters: span.clone().with_range(func_def.parameters.range),
+            return_type: return_type_range.map(|range| span.clone().with_range(range)),
+        })
+    }
+}
+
+/// Representation of a function definition in the AST, along with any previous overloads of the
+/// function. Each overload can be separately generic or not, and each generic overload uses
+/// distinct typevars.
+///
+/// # Ordering
+/// Ordering is based on the function's id assigned by salsa and not on the function literal's
+/// values. The id may change between runs, or when the function literal was garbage collected and
+/// recreated.
+#[salsa::interned(debug)]
+#[derive(PartialOrd, Ord)]
+pub struct FunctionLiteral<'db> {
+    pub(crate) last_definition: OverloadLiteral<'db>,
+
+    /// The inherited generic context, if this function is a class method being used to infer the
+    /// specialization of its generic class. If any of the method's overloads are themselves
+    /// generic, this is in addition to those per-overload generic contexts.
+    inherited_generic_context: Option<GenericContext<'db>>,
+}
+
+#[salsa::tracked]
+impl<'db> FunctionLiteral<'db> {
+    fn with_inherited_generic_context(
+        self,
+        db: &'db dyn Db,
+        inherited_generic_context: GenericContext<'db>,
+    ) -> Self {
+        // A function cannot inherit more than one generic context from its containing class.
+        debug_assert!(self.inherited_generic_context(db).is_none());
+        Self::new(
+            db,
+            self.last_definition(db),
+            Some(inherited_generic_context),
+        )
+    }
+
+    fn update_last_definition(
```

Comment:
nit: I'd prefer to just inline this as I think it's only being used once? Or, if you have any other use cases in mind, feel free to keep it.

Reply:
Yep good call